[cms-v2]Inital work for cms refactoring

fernandoabolafio opened the pull request:
This PR is the first of many to come to accomplish what is described in #1604.

This PR contains the following features:

- [x] Authentication 
     - [x] Login
     - [x] Signup
     - [x] Reset password
     - [x] Resend verification email
- [x] Invoice
    - [x] User List
    - [x] Admin List
    - [x] New
    - [x] Admin actions (approve, reject, dispute)
    - [x] Comments
- [x] User/Contractor 
    - [x] Invite
    - [x] Manage (type, domain, supervisors)

This is a preview of how the main screens are currently looking with the refactoring:

<img width="1038" alt="Screenshot 2019-12-03 at 16 22 50" src="https://user-images.githubusercontent.com/14864439/70069197-4c4e8900-15e9-11ea-8e74-d31531313a12.png">
<img width="1028" alt="Screenshot 2019-12-03 at 16 25 58" src="https://user-images.githubusercontent.com/14864439/70069393-9df71380-15e9-11ea-8860-f2e09e5ec7ab.png">

Depends on https://github.com/decred/politeia/pull/1049
Depends on https://github.com/decred/politeia/pull/1089

**obs:** This PR was firstly open on https://github.com/decred/politeiagui/pull/1605. It was mistakenly merged in the branch cmsv2 so I had to reopen it with a fresh PR. Sorry for the confusion.

thi4go commented on src/reducers/models/invoices.js:
```diff
@@ -0,0 +1,64 @@
+          },
+          [act.RECEIVE_USER_INVOICES]: () =>
+            onReceiveInvoices(state, action.payload.invoices),
+          [act.RECEIVE_ADMIN_INVOICES]: () =>
```
if `RECEIVE_USER_INVOICES` does the same as `RECEIVE_ADMIN_INVOICES` to the state branch, do we need both of them?

fernandoabolafio replied:
Yeah, I think we can merge both into a more generic one like `RECEIVE_INVOICES`. Good catch.

fernandoabolafio replied:
After thinking about it I will keep them as it is just for the sake of debugging and to keep the consistency where different actions have different actions representing what kind of data they are requesting/receiving. Once we start looking deeper to our actions structure we may want to change that.

thi4go commented on src/actions/app.js:
```diff
@@ -54,28 +54,26 @@ export const onSetReplyParent = (parentId = TOP_LEVEL_COMMENT_PARENTID) => (
+  exchangerate,
+  files
+}) => (dispatch, getState) => {
+  const email = sel.loggedInAsEmail(getState());
```
I think we can transition to use the selectors from the new user state branch model. i.e: `currentUserEmail`

thi4go commented on src/actions/app.js:
```diff
@@ -54,28 +54,26 @@ export const onSetReplyParent = (parentId = TOP_LEVEL_COMMENT_PARENTID) => (
+  files
+}) => (dispatch, getState) => {
+  const email = sel.loggedInAsEmail(getState());
+  const id = sel.userid(getState());
```
Same here, use more recent selector `currentUserID`

thi4go commented on src/actions/app.js:
```diff
@@ -54,28 +54,26 @@ export const onSetReplyParent = (parentId = TOP_LEVEL_COMMENT_PARENTID) => (
+}) => (dispatch, getState) => {
+  const email = sel.loggedInAsEmail(getState());
+  const id = sel.userid(getState());
+  const username = sel.loggedInAsUsername(getState());
```
Same as above. I think we can apply this transition throughout the code on the remaining old selectors (correct me if I'm wrong, mate)

fernandoabolafio commented on src/containers/User/Detail/ManageContractor/AdminView.jsx:
```diff
@@ -0,0 +1,73 @@
+const ManageContractor = ({}) => {
+  return (
+    <Card className={classNames("container", "margin-bottom-m")}>
+      <Formik initialValues={{ domain: 1, type: 1, users: [] }}>
```
yes, this is an ongoing work. Sorry for not pointing this out in the code.

fernandoabolafio commented:
@thi4go I addressed all of your review comments. Regarding:

> I get a Error: The provided signature was invalid. from the api when I try to do any admin action. Need to check this again on my linux env, but would be nice if you tested it there as well. 

This won't happen anymore once decred/politeia#1089 is in.

thi4go commented:
Hey @fernandoabolafio, looking very good on my end, tested the listed features and they are all working fine. Just those two minor comments:

- The description needs to reset after each new row
- I'm not sure how this is being done, but I wasn't able to set a supervisor to a user from my admin account.

Another than this, the PR is LGTM.

fernandoabolafio commented:
@thi4go can you describe what was happening when you tried to set a supervisor? Perhaps a small video or a gif would be nice.

> I'm not sure how this is being done, but I wasn't able to set a supervisor to a user from my admin account.

thi4go commented:
@fernandoabolafio no users were being listed as possible supervisors for my users, maybe because I only had one admin account, or a incorrect user status to make it possible.

alexlyp commented on src/pages/RoutesCMS.jsx:
```diff
@@ -0,0 +1,127 @@
+    if (location.pathname !== to) {
+      history.push({ pathname: to, search: location.search });
+    }
+  }, [history, location.pathname, to]);
```
I see that as well.

alexlyp commented:
I'm getting an error `TypeError: supportedChars is undefined` when going to the Account page.

fernandoabolafio commented:
@victorgcramos regarding your comments:

> 1. The mobile view needs some left padding for dropdown menu options, see:

I centralized the content and the dropdown should be fine now.

> 2. so, on mobile view, at least for Safari 13, I was not able to edit the subdomain field. All other fields are okay.

Let's create an issue for that aside with some other UX improvements for the submission form. 
I don't want to keep growing this PR indefinitely for UX tweaks. 

>  3. The manage user tab should be only available for admins.

Why? This tab shows a different view for non-admins (fields aren't editable), so the user can inspect his own attributes for DCC. 

> 4. One quick observation: if the user is not allowed to skip lines, why do we need this multi-line editor field? After filling my invoice description, I noticed that I cannot skip lines. It would be better if I got the error message before clicking "save" on "Edit description" modal. What do you think?

Good catch! I replaced the multiline editor by a simple text input so it shouldn't be a problem anymore.

fernandoabolafio commented:
@alexlyp regarding
> I'm getting an error `TypeError: supportedChars is undefined` when going to the Account page.
> 
> `TypeError: supportedChars is undefined buildUsernameRegex src/github.com/decred/politeiagui/src/containers/User/Detail/validation.js:23`

It was happening before the changes applied by https://github.com/decred/politeia/pull/1049. Are you running politeia with this commit included?

victorgcramos commented:
> Why? This tab shows a different view for non-admins (fields aren't editable), so the user can inspect his own attributes for DCC.

I just said that because of the specs from https://github.com/decred/politeiagui/issues/1470. Nvm!

alexlyp commented:
> It was happening before the changes applied by [decred/politeia#1049](https://github.com/decred/politeia/pull/1049). Are you running politeia with this commit included?

Yah looks like I was running an older version of piwww, all good now. @fernandoabolafio

tiagoalvesdulce commented:
All good here. Great job!

alexlyp commented:
Looks good to me as well!  

@fernandoabolafio I suppose we can merge this then @victorgcramos can start to work on integrating the recent invoice work (which is up for PR) and the DCC work as well.  But if you have any ideas to ease that process I'm interested to hear your ideas.

fernandoabolafio commented:
@alexlyp I think that's the general plan. Merge this and then add all the other features. 
I think is perfect that @victorgcramos works with the features he has been working already. Which is mostly the new things we're adding. @victorgcramos can you organize the work to be done and create issues for that? Such as Invoice improvements and DCC?

And the others can work on adding the existent features that I couldn't add yet. Such as payouts, invoice editing and so on. I created issues for those. 

At some point, we will need to halt the work in the "old code" to be able to deliver the new web-app. I would rather do it sooner than later. What do you think?

fernandoabolafio commented:
Let's move this discussion to https://github.com/decred/politeiagui/issues/1685.
